### PR TITLE
Introduce application credentials auth support

### PR DESCRIPTION
**Pull request by kayrus**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -110,6 +110,7 @@ environment variable:
 
 ```yaml
 password: { fromEnv: ENVIRONMENT_VARIABLE }
+application_credential_secret: { fromEnv: ENVIRONMENT_VARIABLE }
```
Comment (majewsky) [inline README.md]: Please make that into a full example (and also explain `application_credential_{id,name}`), separate from the username/password-based examples, and link to the full config file like the other examples do.

File: pkg/objects/swift.go [hunk 2]
```diff
@@ -57,16 +62,23 @@ type SwiftLocation struct {
 	FileExists map[string]bool `yaml:"-"`
 }
 
-func (s SwiftLocation) cacheKey() string {
-	return strings.Join([]string{
+func (s SwiftLocation) cacheKey(name string) string {
```
Comment (majewsky) [inline pkg/objects/swift.go]: Why is `cacheKey` taking the `name` argument here? If I understand correctly, this breaks the entire purpose of the function, since now it cannot reuse connections across `SwiftLocation` instances anymore.

File: pkg/objects/swift.go [hunk 3]
```diff
@@ -104,40 +134,69 @@ func (s SwiftLocation) Validate(name string) []error {
...
-			IdentityEndpoint: s.AuthURL,
-			Username:         s.UserName,
-			DomainName:       s.UserDomainName,
-			Password:         string(s.Password),
-			Scope: &gophercloud.AuthScope{
-				ProjectName: s.ProjectName,
-				DomainName:  s.ProjectDomainName,
-			},
-			AllowReauth: true,
+		authInfo := &clientconfig.AuthInfo{
```
Comment (majewsky) [inline pkg/objects/swift.go]: Why this change from `gophercloud.AuthOptions` to `clientconfig.AuthInfo`? AuthOptions already supports application credentials.

Comment (kayrus) [reply pkg/objects/swift.go]: application credentials don't require scope, I don't want to introduce yet another `if then else` logic, it's better to reuse the existing package.

File: pkg/objects/swift.go (hunk 2 above)
Comment (kayrus) [reply pkg/objects/swift.go]: in case when you need a debug, you don't care about the efficiency, but want to see what happens and why. without the name parameter you won't be able to identify where are the source client logs and where  the target client logs.

File: pkg/objects/swift.go (hunk 3 above)
Comment (majewsky) [reply pkg/objects/swift.go]: I think it's better to use a small if-then-else than to pull in another entire module.

**Comment by majewsky** [comment]
Also, you may want to rebase onto current master. I switched to Go modules over there and also bumped all dependencies in the process.

File: pkg/objects/swift.go (hunk 3 above)
Comment (kayrus) [reply pkg/objects/swift.go]: this module already handles all the possible auth conditions and well tested. if new auth case is added, it is easier just to bump the dependency or set an extra parameter.

File: README.md (hunk 1 above)
Comment (kayrus) [reply README.md]: done

File: pkg/objects/swift.go (hunk 3 above)
Comment (majewsky) [reply pkg/objects/swift.go]: All that clientconfig does is produce an AuthOptions struct anyway. As a big proponent of the YAGNI principle (You Ain't Gonna Need It), I reverted this change before merging to keep the implementation simple.

**Comment by majewsky** [comment]
I merged this branch manually with some additional amendments to the documentation and changelog. You might especially be interested in https://github.com/sapcc/swift-http-import/commit/746c8774948a7a0d2020fd378dcd250ea7f171dc: Creating new `http.Client` instances from scratch breaks several default behaviors of the HTTP client library, e.g. respecting the `http_proxy` etc. environment variables. (This is also something that Gophercloud gets wrong all the time.)

**Comment by kayrus** [comment]
Unfortunately the https://github.com/sapcc/swift-http-import/commit/746c8774948a7a0d2020fd378dcd250ea7f171dc doesn't work, when `DEBUG=true` env is set.

```
Rt RoundTripper is nil, aborting
2019/12/10 14:50:18 ERROR: could not list objects in Swift at mirror2/ubuntu-repos/: not connected to Swift
```

**Comment by majewsky** [comment]
Gah, `client.Roundtripper` does not follow the conventions of `http.Client`. Fixed in https://github.com/sapcc/swift-http-import/commit/529a99ea4ba43dfe53180275094b70581be66e5a.

**Comment by kayrus** [comment]
@majewsky thanks, auth works fine now.
